cmd/kola: add multiple etcd discovery tests and better vm cleanup

pbx0 opened the pull request:
Add multiple discovery tests for different etcd versions and configurations. Requires image build with both versions: export ETCD_PROTOCOLS="1 2"

/cc @marineam

marineam commented on cmd/kola/etcd.go:
```diff
@@ -48,51 +48,74 @@ func runEtcd(args []string) int {
+	var machines []platform.Machine
+TestLoop:
 	for _, t := range etcdtests.Tests {
+		// kill any machines from the last test instead of waiting for
```
How about splitting the body of this loop into a new function. That way you can leave it to a `defer cluster.Destroy()` and clean up a lot of the complexity of this code in general.

marineam commented on cmd/kola/etcdtests/etcdDiscovery.go:
```diff
@@ -22,29 +24,40 @@ func etcdDiscovery(cluster platform.Cluster) error {
 		// start etcd instance
 		etcdStart := "sudo systemctl start etcd.service"
-		_, err := m.SSH(etcdStart)
+		//etcdStart := "etcd version"
```
left over debugging?

marineam commented on platform/local/cluster.go:
```diff
@@ -102,7 +102,7 @@ func (lc *LocalCluster) GetDiscoveryURL(size int) (string, error) {
 
 	body := strings.NewReader(url.Values{"value": {strconv.Itoa(size)}}.Encode())
-	req, err := http.NewRequest("PUT", URL+"_config/size", body)
+	req, err := http.NewRequest("PUT", URL+"/_config/size", body)
```
Use `path.Join` should be used to avoid similar errors in the future.

pbx0 replied:
path.Join mangles the double-slash in the URL though.

marineam replied:
Ah, yeah, and then more complicated forms of URL construction are likely harder to read. So this is fine then.

philips replied:
Uppercase variable names are generally reserved for exported variables, not local function variables, please call it something like `base` or something.

marineam commented on cmd/kola/etcdtests/etcdDiscovery.go:
```diff
@@ -22,29 +24,40 @@ func etcdDiscovery(cluster platform.Cluster) error {
+		}
+		members = strings.Split(string(b), ",")
+		// etcd 2.0 listens on 2 ports for clients and returne 2*csize
+		if len(members) == csize || len(members) == csize*2 {
```
It would be more robust to compare the list of members to the list of machines.

pbx0 replied:
Curling `/v2/machines` seemed to be the only version agnostic way to easily see that the cluster had formed correctly, but is slightly ugly in that 2.0.x normally returns a list that is double the cluster size.  curling `/v2/members` is 2.0.x only.

I'll check that `etcdctl cluster-health` returns 3 healthy members instead.

marineam replied:
I was thinking along the lines of parsing the output of machines to verify the hosts you expected were there. The check could just ignore duplicated hosts on different ports because it is skimming for IP or host id.

marineam commented:
Overall this is better but I think it deserves another cleanup pass. As a general trend loops seem more complicated than they need to be. Breaking up into functions more should help.

pbx0 commented:
PTAL

I did some cleanup in etcd.go and etcdDiscovery.go Checking for cluster success now utilizes `etcdctl cluster-health` which should be more robust.

philips commented on cmd/kola/etcd.go:
```diff
@@ -33,7 +33,7 @@ var cmdEtcd = &cli.Command{
 Work in progress: the code sets up an etcd cluster and then calls
 integration tests to see how etcd interacts with new CoreOS images.
-Currently discovery using another etcd cluster is supported for 0.4.7
+Currently discovery using an embedded etcd server is supported.
```
This sentence doesn't parse for me. I think it is OK if "is supported" is removed and a comma is added after "Currently".

philips commented on cmd/kola/etcd.go:
```diff
@@ -48,51 +48,53 @@ func runEtcd(args []string) int {
 		}
+		fmt.Printf("Test %v ran successfully\n", t.Name)
+	}
+	fmt.Fprintf(os.Stderr, "Etcd tests ran successfully!\n")
```
Etcd should always be lowercase

philips commented on cmd/kola/etcdtests/etcdDiscovery.go:
```diff
@@ -22,30 +22,47 @@ func etcdDiscovery(cluster platform.Cluster) error {
 			return fmt.Errorf("SSH cmd to %v failed: %s", m.IP(), err)
 		}
-		fmt.Fprintf(os.Stderr, "etcd instance%d started\n", i)
+		fmt.Fprintf(os.Stderr, "etcd instance%d started%s\n", i, b)
```
space after started?

pbx0 commented:
PTAL

Fixed comments from @philips and squashed commits into one. The second commit re-fixed the vm cleanup loop code (ugly) in the first commit which doesn't need to be in the history.

marineam commented on cmd/kola/etcd.go:
```diff
@@ -48,51 +49,53 @@ func runEtcd(args []string) int {
-				fmt.Fprintf(os.Stderr, "Destroy failed: %v\n", err)
-				return 1
+		defer func() {
+			if err := m.Destroy(); err != nil {
```
cluster.Destroy cleans up machines so this shouldn't be required.

marineam commented:
Code is looking much better now, once you address the two remaining comments this should be good to go.

pbx0 commented:
Fixed up last two comments and merging
